language/stdlib/modules/validator_config_2.move

```diff
@@ -0,0 +1,97 @@
+address 0x0:
+
+module ValidatorConfig2 {
+    use 0x0::LibraAccount;
+    use 0x0::Transaction;
+
+    struct Config {
+        consensus_pubkey: vector<u8>,
+    }
+
+    // A current or prospective validator should publish one of these under their address.
+    // If delegation is enabled, only the delegated_address can change the config.
+    // If delegation is disabled, only the owner of this resource can change the config.
+    // The entity that can change the config is called a "validator operator".
+    // The owner of this resource can enable delegation and set delegated address.
+    // The owner can also disable delegation at any time.
+    resource struct T {
+        config: Config,
+        delegation_enabled: bool,
+        delegated_address: address,
+    }
+
+    // TODO(valerini): add events here
+
+    // Returns true if addr has a published ValidatorConfig::T resource
+    public fun has(addr: address): bool { exists<T>(addr) }
+
+    // Getters for Config
+    public fun get_config(addr: address): Config acquires T { *&borrow_global<T>(addr).config }
+    public fun get_consensus_pubkey(config_ref: &Config): vector<u8> { *&config_ref.consensus_pubkey }
+
+    // Returns the address of the entity who is now in charge of managing the config.
+    public fun get_validator_operator_address(addr: address): address acquires T {
+        let t_ref = borrow_global<T>(addr);
+        if (t_ref.delegation_enabled) {
+            return *&t_ref.delegated_address
+        } else {
+            return *&addr
+        }
+    }
+
+    // Register the transaction sender as a candidate validator by creating a ValidatorConfig
+    // resource under their account. Note that only one such resource can be instantiated under an account.
+    public fun initialize(
+        consensus_pubkey: vector<u8>,) {
+
+        move_to_sender<T>(
+            T {
+                config: Config {
+                    consensus_pubkey: consensus_pubkey,
+                },
+                delegation_enabled: false,
+                delegated_address: 0x0
+            }
+        );
+    }
+
+    // If the sender of the transaction has a ValidatorConfig::T resource,
+    // this function will delegate management of the ValidatorConfig::T resource to a delegated_address.
+    public fun set_delegated_address(delegated_address: address) acquires T {
```

Comment:
super nit: do we want to refer to these as accounts or addresses? I personally prefer the former.

Reply:
I think "account" makes sense here because the (enforced) intention is to choose an address at which an account exists.

Reply:
Sure! I have no preference - changed to account:
`delegated_address field -> delegated_account`,
`set_delegated_address -> set_delegated_account`,
`remove_delegated_address -> remove_delegated_account`.